pym/bob/cmds/build.py

```diff
@@ -1068,3 +1069,107 @@ def doClean(argv, bobRoot):
+
+    # Parse args
+    args = parser.parse_args(argv)
+    if args.sandbox == None:
+        args.sandbox = not args.dev
+
+    # Process defines
+    defines = {}
+    for define in args.defines:
+        d = define.split("=")
+        if len(d) == 1:
+            defines[d[0]] = ""
+        elif len(d) == 2:
+            defines[d[0]] = d[1]
+        else:
+            parser.error("Malformed define: "+define)
+
+    # Process the recipes
+    recipes = RecipeSet()
+    recipes.defineHook('releaseNameFormatter', LocalBuilder.releaseNameFormatter)
+    recipes.defineHook('developNameFormatter', LocalBuilder.developNameFormatter)
+    recipes.defineHook('developNamePersister', LocalBuilder.developNamePersister)
+    recipes.setConfigFiles(args.configFile)
+    recipes.parse()
+
+    # State variables in a class
+    class State:
+        def __init__(self):
+            self.packageText = ''
+            self.showPackage = True
+        def appendText(self, what):
+            self.packageText += what
+        def appendStep(self, step):
+            dir = step.getWorkspacePath()
+            if step.isValid() and (dir is not None) and os.path.isdir(dir):
+                self.packageText += dir
+            else:
+                self.showPackage = False
+        def print(self):
+            if (self.showPackage):
+                print(self.packageText)
+
+    if args.dev:
+        # Develop names are stable. All we need to do is to replicate build's algorithm,
+        # and when we produce a name, check whether it exists.
+        nameFormatter = recipes.getHook('developNameFormatter')
+        developPersister = recipes.getHook('developNamePersister')
+        nameFormatter = developPersister(nameFormatter)
+    else:
+        # Release names are taken from persistence.
+        nameFormatter = LocalBuilder.releaseNameInterrogator
+    nameFormatter = LocalBuilder.makeRunnable(nameFormatter)
+
+    # Find roots
+    roots = recipes.generatePackages(nameFormatter, defines, args.sandbox)
+    if args.dev:
+        touch(sorted(roots.values(), key=lambda p: p.getName()))
+
+    # Loop through packages
+    for p in args.packages:
+        # Format this package.
+        # Only show the package if all of the requested directory names are present
+        package = walkPackagePath(roots, p)
+        state = State()
+        for (text, var, spec, conversion) in Formatter().parse(args.f):
+            state.appendText(text)
+            if var is None:
+                pass
+            elif var == 'name':
+                state.appendText(p)
+            elif var == 'src':
+                state.appendStep(package.getCheckoutStep())
+            elif var == 'build':
+                state.appendStep(package.getBuildStep())
+            elif var == 'dist':
+                state.appendStep(package.getPackageStep())
+            else:
+                raise KeyError("Unknown field '{" + var + "}'")
```

Comment:
You should raise a ParseError here. Only exceptions derived from bob.errors.BobError are caught gracefully:

```
jkloetzke@tuxedo:~/src/bob/playground/recipes$ bob query-path -f "{name} {foobar}" foo/bar 
An internal Exception has occured. This should not have happenend.
Please open an issue at https://github.com/BobBuildTool/bob with the following backtrace:
Traceback (most recent call last):
  File "/home/jkloetzke/src/bob/bob/pym/bob/scripts.py", line 103, in bob
    availableCommands[verb][1](argv, bobRoot)
  File "/home/jkloetzke/src/bob/bob/pym/bob/scripts.py", line 59, in __querypath
    doQueryPath(*args, **kwargs)
  File "/home/jkloetzke/src/bob/bob/pym/bob/cmds/build.py", line 1172, in doQueryPath
    raise KeyError("Unknown field '{" + var + "}'")
KeyError: "Unknown field '{foobar}'"
```